clang/include/clang/Tooling/DependencyScanning/DependencyScanningFilesystem.h

```diff
@@ -226,9 +247,28 @@ class DependencyScanningFilesystemLocalCache {
   insertEntryForFilename(StringRef Filename,
                          const CachedFileSystemEntry &Entry) {
     assert(llvm::sys::path::is_absolute_gnu(Filename));
-    const auto *InsertedEntry = Cache.insert({Filename, &Entry}).first->second;
-    assert(InsertedEntry == &Entry && "entry already present");
-    return *InsertedEntry;
+    assert(Cache[Filename].first == nullptr && "entry already present");
+    Cache[Filename].first = &Entry;
+    return Entry;
```

Comment:
Nit: I think the original way this was written is slightly better, because in a no-asserts build if this is ever hit, it returns the entry that's actually in the map and in an asserts build it is cheaper to only lookup in the map once instead of twice. I would just keep the old way but insert a pair(&Entry, nullptr)

Reply:
The complication is that the cache entry may now already exist, but will only have the real path populated. I guess we can get close to the existing behavior with something like this:

```
    auto [It, Inserted] = Cache.insert({Filename, {&Entry, nullptr}});
    auto &[CachedEntry, CachedRealPath] = It->getValue();
    if (!Inserted) {
      // The file is already present in the local cache. If we got here, it only
      // contains the real path. Let's make sure the entry is populated too.
      assert((!CachedEntry && CachedRealPath) && "entry already present");
      CachedEntry = &Entry;
    }
    return *CachedEntry;
```